pypeit/core/datacube.py

```diff
@@ -989,6 +1029,337 @@ def compute_weights(all_ra, all_dec, all_wave, all_sci, all_ivar, all_idx, white
+        # Generate a spline between spectral pixel position and wavelength
+        yspl = tilts[wpix]*(slits.nspec - 1)
+        wspl = all_wave[this_sl]
+        asrt = np.argsort(yspl)
+        wave_spl = interp1d(yspl[asrt], wspl[asrt], kind='linear', bounds_error=False, fill_value='extrapolate')
+        for xx in range(subsample):
+            for yy in range(subsample):
+                # Calculate the tranformation from detector pixels to voxels
+                spatpos = astrom_trans.transform(sl, wpix[1] + ssamp_offs[xx], wpix[0] + ssamp_offs[yy])
+                # TODO :: The tilts in the following line is evaluated at the pixel location, not the subsampled pixel location
+                # A simple fix is implemented for the spectral direction, but this is not so straightforward for the spatial direction
+                # Probably, the correction in the spatial direction is so tiny, that this doesn't matter...
+                specpos = (wave_spl(tilts[wpix]*(slits.nspec - 1) + ssamp_offs[yy]) - wave0) / wave_delta
+                # Now assemble this position of the datacube
+                pix_coord = np.column_stack((slitID, spatpos, specpos))
+                tmp_dc, _ = np.histogramdd(pix_coord, bins=bins, weights=all_sci[this_sl] * all_wght_subsmp[this_sl])
+                tmp_vr, _ = np.histogramdd(pix_coord, bins=bins, weights=all_var[this_sl] * all_wght_subsmp[this_sl]**2)
+                tmp_nm, _ = np.histogramdd(pix_coord, bins=bins, weights=all_wght_subsmp[this_sl])
+                if debug:
+                    tmp_rsd, _ = np.histogramdd(pix_coord, bins=bins, weights=all_sci[this_sl] * np.sqrt(all_ivar[this_sl]))
+                    residcube += tmp_rsd
+                datacube += tmp_dc
+                varcube += tmp_vr
+                normcube += tmp_nm
+    # Normalise the datacube and variance cube
+    nc_inverse = utils.inverse(normcube)
+    datacube *= nc_inverse
+    varcube *= nc_inverse**2
+    if debug: residcube *= nc_inverse
+
+    # Prepare the header, and add the unit of flux to the header
+    hdr = output_wcs.to_header()
+    if fluxcal:
+        hdr['FLUXUNIT'] = (PYPEIT_FLUX_SCALE, "Flux units -- erg/s/cm^2/Angstrom/arcsec^2")
+    else:
+        hdr['FLUXUNIT'] = (1, "Flux units -- counts/s/Angstrom/arcsec^2")
+
+    # Write out the datacube
+    msgs.info("Saving datacube as: {0:s}".format(outfile))
+    final_cube = DataCube(datacube.T, varcube.T, specname, blaze_wave, blaze_spec, sensfunc=sensfunc, fluxed=fluxcal)
+    final_cube.to_file(outfile, hdr=hdr, overwrite=overwrite)
+
+    # Save a residuals cube, if requested
+    if debug:
+        outfile_resid = outfile.replace(".fits", "_resid.fits")
+        msgs.info("Saving residuals datacube as: {0:s}".format(outfile_resid))
+        hdu = fits.PrimaryHDU(residcube.T, header=hdr)
+        hdu.writeto(outfile_resid, overwrite=overwrite)
+
+
 def generate_cube_ngp(outfile, hdr, all_sci, all_ivar, all_wghts, pix_coord, bins,
```

Comment:
Given the overlap with the new subsample function, do we still need this ngp function?

Reply:
Yes... for the moment... This function is used when combining multiple cubes together. The subsample function is only used for generating a NGP cube when converting single spec2d files into a single datacube. I can't find a resource friendly way of doing this to multiple frames, yet.